### ISSUE TITLE
Error cross-compiling for ARMv7

### ISSUE THREAD

**Issue by martinboers** [objectcomputing/OpenDDS#1294]
Hello,
I am having trouble cross-compiling OpenDDS for an ARMv7 platform.
I have installed the SDK for the target, and I am able to cross-compile plenty of open-source libraries using this SDK.
For OpenDDS, I have run the following:
```sh
./configure --target=linux-cross --target-compiler=/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin/arm-pxc-linux-gnueabi/arm-pxc-linux-gnueabi-g++ --no-tests
```
... which results in the following configuration:
```sh
$ tools/scripts/show_build_config.pl
Host Build:
================ setenv.sh ================
# OpenDDS configure script: ./configure --target=linux-cross --target-compiler=/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin/arm-pxc-linux-gnueabi/arm-pxc-linux-gnueabi-g++ --no-tests
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}
export PATH=${PATH}
export MPC_ROOT=/home/tcs-user/Documents/projects/OpenDDS/ACE_wrappers/MPC
export CIAO_ROOT=unused
export DDS_ROOT=/home/tcs-user/Documents/projects/OpenDDS/build/host
export DANCE_ROOT=unused
export TAO_ROOT=/home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/TAO
export ACE_ROOT=/home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers
================ user_macros.GNU ================
================ platform_macros.GNU ================
static_libs_only = 1
include $(ACE_ROOT)/include/makeinclude/platform_linux.GNU
================ config.h ================
#include "ace/config-linux.h"
Target Build:
================ setenv.sh ================
# OpenDDS configure script: ./configure --target=linux-cross --target-compiler=/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin/arm-pxc-linux-gnueabi/arm-pxc-linux-gnueabi-g++ --no-tests
export PATH=${PATH}:/home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/bin:/home/tcs-user/Documents/projects/OpenDDS/build/host/bin
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/home/tcs-user/Documents/projects/OpenDDS/build/target/ACE_wrappers/lib:/home/tcs-user/Documents/projects/OpenDDS/build/target/lib
export MPC_ROOT=/home/tcs-user/Documents/projects/OpenDDS/ACE_wrappers/MPC
export DDS_ROOT=/home/tcs-user/Documents/projects/OpenDDS/build/target
export CIAO_ROOT=unused
export HOST_DDS=/home/tcs-user/Documents/projects/OpenDDS/build/host
export HOST_ACE=/home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers
export DANCE_ROOT=unused
export ACE_ROOT=/home/tcs-user/Documents/projects/OpenDDS/build/target/ACE_wrappers
export TAO_ROOT=/home/tcs-user/Documents/projects/OpenDDS/build/target/ACE_wrappers/TAO
================ user_macros.GNU ================
OPENDDS_IDL = $(HOST_DDS)/bin/opendds_idl
OPENDDS_IDL_DEP = $(OPENDDS_IDL)$(HOST_EXE_EXT)
IDL2JNI = $(HOST_DDS)/bin/idl2jni
IDL2JNI_DEP = $(IDL2JNI)$(HOST_EXE_EXT)
CROSS-COMPILE = 1
================ default.features ================
no_cxx11=0
cross_compile=1
================ platform_macros.GNU ================
optimize = 0
TAO_IDL = $(HOST_ACE)/bin/tao_idl
TAO_IDL_DEP = $(TAO_IDL)$(HOST_EXE_EXT)
TAO_IDL_PREPROCESSOR = g++
TAO_IDLFLAGS += -g $(HOST_ACE)/bin/ace_gperf
build_tao_idl_be = 0
CROSS_COMPILE = /opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin/arm-pxc-linux-gnueabi/arm-pxc-linux-gnueabi-
include $(ACE_ROOT)/include/makeinclude/platform_linux.GNU
================ config.h ================
#include "ace/config-linux.h"
```
When I run `make`, I eventally end up with the following error:
```sh
arm-pxc-linux-gnueabi-g++ -Wnon-virtual-dtor -O3 -ggdb -pthread -fno-strict-aliasing -Wall -W -Wpointer-arith -pipe  -D_GNU_SOURCE -DACE_HAS_CUSTOM_EXPORT_MACROS=0   -I/home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers -D__ACE_INLINE__ -I.. -DACE_AS_STATIC_LIBS  -c  -o arm/.obj/Local_Name_Space.o Local_Name_Space.cpp
In file included from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/config-linux.h:24,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/config.h:1,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/config-macros.h:24,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/config-lite.h:24,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/ACE_export.h:11,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/Malloc.h:18,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/Malloc_T.h:18,
                 from /home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/Local_Name_Space.h:20,
                 from Local_Name_Space.cpp:3:
/home/tcs-user/Documents/projects/OpenDDS/build/host/ACE_wrappers/ace/config-linux-common.h:18:10: fatal error: features.h: No such file or directory
 #include <features.h>
          ^~~~~~~~~~~~
compilation terminated.
```
It looks like the ACE/TAO build does not know where to find the include files in the SDK, which makes sense because this information was not provided as part of the `configure` step.

The SDK comes with an environment setup file, which I normally `source` before cross-compiling, but which I have not used in this case (it causes the OpenDDS `configure` step to fail). Here are the contents of the SDK environment setup file:
```sh
# Check for LD_LIBRARY_PATH being set, which can break SDK and generally is a bad practice
# http://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html#AEN80
# http://xahlee.info/UnixResource_dir/_/ldpath.html
# Only disable this check if you are absolutely know what you are doing!
if [ ! -z "$LD_LIBRARY_PATH" ]; then
    echo "Your environment is misconfigured, you probably need to 'unset LD_LIBRARY_PATH'"
    echo "but please check why this was set in the first place and that it's safe to unset."
    echo "The SDK will not operate correctly in most cases when LD_LIBRARY_PATH is set."
    echo "For more references see:"
    echo "  http://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html#AEN80"
    echo "  http://xahlee.info/UnixResource_dir/_/ldpath.html"
    return 1
fi
export SDKTARGETSYSROOT=/opt/pxc/sdk/AXCF2152/2019.6/sysroots/cortexa9t2hf-neon-pxc-linux-gnueabi
export PATH=/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin:/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/sbin:/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/bin:/
opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/sbin:/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin/../x86_64-pokysdk-linux/bin:/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-
linux/usr/bin/arm-pxc-linux-gnueabi:/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/bin/arm-pxc-linux-musl:$PATH
export PKG_CONFIG_SYSROOT_DIR=$SDKTARGETSYSROOT
export PKG_CONFIG_PATH=$SDKTARGETSYSROOT/usr/lib/pkgconfig:$SDKTARGETSYSROOT/usr/share/pkgconfig
export CONFIG_SITE=/opt/pxc/sdk/AXCF2152/2019.6/site-config-cortexa9t2hf-neon-pxc-linux-gnueabi
export OECORE_NATIVE_SYSROOT="/opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux"
export OECORE_TARGET_SYSROOT="$SDKTARGETSYSROOT"
export OECORE_ACLOCAL_OPTS="-I /opt/pxc/sdk/AXCF2152/2019.6/sysroots/x86_64-pokysdk-linux/usr/share/aclocal"
export OECORE_BASELIB="lib"
export OECORE_TARGET_ARCH="arm"
export OECORE_TARGET_OS="linux-gnueabi"
unset command_not_found_handle
export CC="arm-pxc-linux-gnueabi-gcc  -march=armv7-a -mthumb -mfpu=neon -mfloat-abi=hard -mcpu=cortex-a9 --sysroot=$SDKTARGETSYSROOT"
export CXX="arm-pxc-linux-gnueabi-g++  -march=armv7-a -mthumb -mfpu=neon -mfloat-abi=hard -mcpu=cortex-a9 --sysroot=$SDKTARGETSYSROOT"
export CPP="arm-pxc-linux-gnueabi-gcc -E  -march=armv7-a -mthumb -mfpu=neon -mfloat-abi=hard -mcpu=cortex-a9 --sysroot=$SDKTARGETSYSROOT"
export AS="arm-pxc-linux-gnueabi-as "
export LD="arm-pxc-linux-gnueabi-ld  --sysroot=$SDKTARGETSYSROOT"
export GDB=arm-pxc-linux-gnueabi-gdb
export STRIP=arm-pxc-linux-gnueabi-strip
export RANLIB=arm-pxc-linux-gnueabi-ranlib
export OBJCOPY=arm-pxc-linux-gnueabi-objcopy
export OBJDUMP=arm-pxc-linux-gnueabi-objdump
export AR=arm-pxc-linux-gnueabi-ar
export NM=arm-pxc-linux-gnueabi-nm
export M4=m4
export TARGET_PREFIX=arm-pxc-linux-gnueabi-
export CONFIGURE_FLAGS="--target=arm-pxc-linux-gnueabi --host=arm-pxc-linux-gnueabi --build=x86_64-linux --with-libtool-sysroot=$SDKTARGETSYSROOT"
export CFLAGS=" -O2 -pipe -g -feliminate-unused-debug-types "
export CXXFLAGS=" -O2 -pipe -g -feliminate-unused-debug-types "
export LDFLAGS="-Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed"
export CPPFLAGS=""
export KCFLAGS="--sysroot=$SDKTARGETSYSROOT"
export OECORE_DISTRO_VERSION="2019.6"
export OECORE_SDK_VERSION="2019.6"
export ARCH=arm
export CROSS_COMPILE=arm-pxc-linux-gnueabi-

# Append environment subscripts
if [ -d "$OECORE_TARGET_SYSROOT/environment-setup.d" ]; then
    for envfile in $OECORE_TARGET_SYSROOT/environment-setup.d/*.sh; do
	    . $envfile
    done
fi
if [ -d "$OECORE_NATIVE_SYSROOT/environment-setup.d" ]; then
    for envfile in $OECORE_NATIVE_SYSROOT/environment-setup.d/*.sh; do
	    . $envfile
    done
fi
```
I guess that I need to include some additional build flags, and possibly some environment variables, in the OpenDDS configuration. In particular, the `--sysroot` build flag seems important. I do not see any `configure` parameters that might help, and I am not sure where to add this type of information in the generated OpenDDS configuration files.

Any advice would be welcome.

**Comment by martinboers** [comment]
Solved by building host tools in a separate directory, then building the target using those host tools.

**Comment by mitza-oci** [comment]
If you look back at the compiler command line and the resulting error (see "When I run make,..." in the first post of the Issue), I think the root cause is clear.  Its is running the target compiler during the host build.  This is due to the "environment setup file" setting some commonly used variables like `CXX`.  These must not be set for the OpenDDS build because they interfere with ACE's makefiles.

**Comment by martinboers** [comment]
Yep, I see that the target compiler is being run during the host build. However this is not due to the "environment setup file" - at least not the one that I showed in the original post - because, as I mentioned, I did not use the setup file in this case, so the environment variables were not set to the values shown in that file during the OpenDDS build.

The question remains as to why OpenDDS uses the target compiler instead of the host compiler for the host build, but (as I mentioned) the work-around is to do separate host and target builds.

**Comment by mitza-oci** [comment]
Yes, I saw after I posted that it wasn't necessarily that environment setup file.  The overall point I was trying to make is that the environment is the only thing I know of that could cause this issue.  The way that makefile variables interact with environment variables makes it very prone to this sort of problem, we've seen it many times.  Before running `make` for OpenDDS, take a look at the shell's environment and see if any of those key variables that ACE's makefiles will use are set.
